File: services/market_data.py
import yfinance as yf
import pandas as pd
import ta
from typing import Dict, List, Optional
from datetime import datetime


class MarketDataService:

    def normalize_symbol(self, symbol: str) -> str:
        symbol = symbol.upper().strip()
        if symbol.startswith("^"):
            return symbol
        if "." not in symbol:
            return f"{symbol}.NS"
        return symbol

    async def get_quote(self, symbol: str) -> Dict:
        try:
            normalized = self.normalize_symbol(symbol)
            ticker = yf.Ticker(normalized)
            hist = ticker.history(period="5d")

            if hist.empty:
                return {"error": f"No data for {symbol}", "symbol": symbol}

            current_price = float(hist["Close"].iloc[-1])
            prev_price = float(hist["Close"].iloc[-2]) if len(hist) > 1 else current_price
            change = current_price - prev_price
            change_pct = (change / prev_price * 100) if prev_price else 0

            volume = hist["Volume"].iloc[-1]

            return {
                "symbol": symbol.upper(),
                "normalized_symbol": normalized,
                "current_price": round(current_price, 2),
                "previous_close": round(prev_price, 2),
                "change": round(change, 2),
                "change_pct": round(change_pct, 2),
                "volume": int(volume) if not pd.isna(volume) else 0,
                "timestamp": datetime.utcnow().isoformat(),
            }
        except Exception as e:
            return {"error": str(e), "symbol": symbol}

    async def get_technical_indicators(self, symbol: str) -> Dict:
        try:
            normalized = self.normalize_symbol(symbol)
            ticker = yf.Ticker(normalized)
            hist = ticker.history(period="1y")

            if hist.empty or len(hist) < 26:
                return {"error": "Insufficient data", "symbol": symbol}

            close = hist["Close"]
            high = hist["High"]
            low = hist["Low"]

            def safe_float(series, idx=-1):
                val = series.iloc[idx]
                return round(float(val), 4) if not pd.isna(val) else None

            rsi = ta.momentum.RSIIndicator(close, window=14).rsi()
            macd_obj = ta.trend.MACD(close)
            sma_20 = ta.trend.SMAIndicator(close, window=20).sma_indicator()
            sma_50 = ta.trend.SMAIndicator(close, window=50).sma_indicator()
            sma_200 = ta.trend.SMAIndicator(close, window=200).sma_indicator()
            ema_20 = ta.trend.EMAIndicator(close, window=20).ema_indicator()
            bb = ta.volatility.BollingerBands(close, window=20)

            current_price = float(close.iloc[-1])

            return {
                "symbol": symbol.upper(),
                "current_price": round(current_price, 2),
                "rsi": safe_float(rsi),
                "macd": {
                    "macd": safe_float(macd_obj.macd()),
                    "signal": safe_float(macd_obj.macd_signal()),
                    "histogram": safe_float(macd_obj.macd_diff()),
                },
                "moving_averages": {
                    "sma_20": safe_float(sma_20),
                    "sma_50": safe_float(sma_50),
                    "sma_200": safe_float(sma_200),
                    "ema_20": safe_float(ema_20),
                },
                "bollinger_bands": {
                    "upper": safe_float(bb.bollinger_hband()),
                    "middle": safe_float(bb.bollinger_mavg()),
                    "lower": safe_float(bb.bollinger_lband()),
                },
                "52w_high": round(float(high.max()), 2),
                "52w_low": round(float(low.min()), 2),
            }
        except Exception as e:
            return {"error": str(e), "symbol": symbol}

    async def get_market_indices(self) -> Dict:
        indices = {
            "NIFTY 50": "^NSEI",
            "SENSEX": "^BSESN",
            "NIFTY BANK": "^NSEBANK",
        }
        results = {}
        for name, symbol in indices.items():
            results[name] = await self.get_quote(symbol)
        return results

    async def get_portfolio_data(self, holdings: List[Dict]) -> List[Dict]:
        enriched = []
        for holding in holdings:
            quote = await self.get_quote(holding["symbol"])
            technical = await self.get_technical_indicators(holding["symbol"])

            current_price = quote.get("current_price", holding["avg_price"])
            invested = holding["quantity"] * holding["avg_price"]
            current_value = holding["quantity"] * current_price
            pnl = current_value - invested
            pnl_pct = (pnl / invested * 100) if invested > 0 else 0

            rsi = technical.get("rsi")
            macd_hist = (technical.get("macd") or {}).get("histogram", 0) or 0
            sma_50 = (technical.get("moving_averages") or {}).get("sma_50")

            enriched.append(
                {
                    **holding,
                    "current_price": current_price,
                    "invested_value": round(invested, 2),
                    "current_value": round(current_value, 2),
                    "pnl": round(pnl, 2),
                    "pnl_pct": round(pnl_pct, 2),
                    "change": quote.get("change", 0),
                    "change_pct": quote.get("change_pct", 0),
                    "rsi": rsi,
                    "rsi_signal": (
                        "oversold" if (rsi or 50) < 30
                        else "overbought" if (rsi or 50) > 70
                        else "neutral"
                    ),
                    "macd_signal": "bullish" if macd_hist > 0 else "bearish",
                    "above_sma_50": current_price > sma_50 if sma_50 else None,
                    "volume": quote.get("volume", 0),
                }
            )
        return enriched

    async def get_historical_data(self, symbol: str, period: str = "1y") -> List[Dict]:
        try:
            normalized = self.normalize_symbol(symbol)
            ticker = yf.Ticker(normalized)
            hist = ticker.history(period=period)

            result = []
            for date, row in hist.iterrows():
                result.append(
                    {
                        "date": date.strftime("%Y-%m-%d"),
                        "open": round(float(row["Open"]), 2),
                        "high": round(float(row["High"]), 2),
                        "low": round(float(row["Low"]), 2),
                        "close": round(float(row["Close"]), 2),
                        "volume": int(row["Volume"]) if not pd.isna(row["Volume"]) else 0,
                    }
                )
            return result
        except Exception:
            return []

File: server/marketData.js
import yahooFinance from 'yahoo-finance2'
import { RSI, MACD, BollingerBands, SMA, EMA } from 'technicalindicators'

yahooFinance.suppressNotices(['yahooSurvey'])

export function normalizeSymbol(symbol) {
  symbol = symbol.toUpperCase().trim()
  if (symbol.startsWith('^')) return symbol
  if (!symbol.includes('.')) return `${symbol}.NS`
  return symbol
}

export async function getQuote(symbol) {
  try {
    const norm = normalizeSymbol(symbol)
    const q = await yahooFinance.quote(norm, {}, { validateResult: false })
    return {
      symbol: symbol.toUpperCase(),
      currentPrice: q.regularMarketPrice ?? 0,
      change: q.regularMarketChange ?? 0,
      changePct: q.regularMarketChangePercent ?? 0,
      previousClose: q.regularMarketPreviousClose ?? 0,
      volume: q.regularMarketVolume ?? 0,
      timestamp: new Date().toISOString()
    }
  } catch (e) {
    return { error: e.message, symbol, currentPrice: 0, change: 0, changePct: 0 }
  }
}

export async function getTechnical(symbol) {
  try {
    const norm = normalizeSymbol(symbol)
    const end = new Date()
    const start = new Date(); start.setFullYear(start.getFullYear() - 1)

    const history = await yahooFinance.historical(norm, {
      period1: start.toISOString().split('T')[0],
      period2: end.toISOString().split('T')[0],
      interval: '1d'
    }, { validateResult: false })

    if (!history?.length || history.length < 30) return { error: 'Insufficient data', symbol }

    const closes = history.map(d => d.close).filter(Boolean)
    const highs  = history.map(d => d.high).filter(Boolean)
    const lows   = history.map(d => d.low).filter(Boolean)
    const last   = arr => arr[arr.length - 1]
    const r      = v => v != null ? Math.round(v * 100) / 100 : null

    const rsiArr  = RSI.calculate({ values: closes, period: 14 })
    const macdArr = MACD.calculate({ values: closes, fastPeriod: 12, slowPeriod: 26, signalPeriod: 9, SimpleMAOscillator: false, SimpleMASignal: false })
    const bbArr   = BollingerBands.calculate({ period: 20, values: closes, stdDev: 2 })
    const sma20   = SMA.calculate({ period: 20, values: closes })
    const sma50   = SMA.calculate({ period: 50, values: closes })
    const sma200  = SMA.calculate({ period: 200, values: closes })
    const ema20   = EMA.calculate({ period: 20, values: closes })

    const cm = last(macdArr) ?? {}
    const cb = last(bbArr) ?? {}

    return {
      symbol: symbol.toUpperCase(),
      currentPrice: r(last(closes)),
      rsi: r(last(rsiArr)),
      macd: { macd: r(cm.MACD), signal: r(cm.signal), histogram: r(cm.histogram) },
      movingAverages: { sma20: r(last(sma20)), sma50: r(last(sma50)), sma200: r(last(sma200)), ema20: r(last(ema20)) },
      bollingerBands: { upper: r(cb.upper), middle: r(cb.middle), lower: r(cb.lower) },
      w52High: r(Math.max(...highs)),
      w52Low: r(Math.min(...lows))
    }
  } catch (e) {
    return { error: e.message, symbol }
  }
}

export async function getMarketIndices() {
  const indices = { 'NIFTY 50': '^NSEI', 'SENSEX': '^BSESN', 'NIFTY BANK': '^NSEBANK' }
  const results = {}
  for (const [name, sym] of Object.entries(indices)) {
    results[name] = await getQuote(sym)
  }
  return results
}

export async function getPortfolioData(holdings) {
  return Promise.all(holdings.map(async h => {
    const [quote, tech] = await Promise.all([getQuote(h.symbol), getTechnical(h.symbol)])
    const price   = quote.currentPrice || h.avg_price
    const invested = h.quantity * h.avg_price
    const current  = h.quantity * price
    const pnl      = current - invested
    const pnlPct   = invested > 0 ? (pnl / invested * 100) : 0
    const rsi      = tech.rsi
    const macdHist = tech.macd?.histogram ?? 0
    const sma50    = tech.movingAverages?.sma50

    return {
      ...h,
      currentPrice:  Math.round(price * 100) / 100,
      investedValue: Math.round(invested * 100) / 100,
      currentValue:  Math.round(current * 100) / 100,
      pnl:     Math.round(pnl * 100) / 100,
      pnlPct:  Math.round(pnlPct * 100) / 100,
      change:    quote.change ?? 0,
      changePct: quote.changePct ?? 0,
      rsi,
      rsiSignal:  !rsi ? 'neutral' : rsi < 30 ? 'oversold' : rsi > 70 ? 'overbought' : 'neutral',
      macdSignal: macdHist > 0 ? 'bullish' : 'bearish',
      aboveSma50: sma50 != null ? price > sma50 : null
    }
  }))
}

export async function getHistory(symbol, period = '1y') {
  try {
    const norm  = normalizeSymbol(symbol)
    const end   = new Date()
    const start = new Date()
    const months = { '1mo': 1, '3mo': 3, '6mo': 6, '1y': 12, '2y': 24 }[period] ?? 12
    start.setMonth(start.getMonth() - months)

    const hist = await yahooFinance.historical(norm, {
      period1: start.toISOString().split('T')[0],
      period2: end.toISOString().split('T')[0],
      interval: '1d'
    }, { validateResult: false })

    return hist.map(d => ({
      date:   d.date.toISOString().split('T')[0],
      open:   d.open,  high: d.high,
      low:    d.low,   close: d.close,
      volume: d.volume
    }))
  } catch { return [] }
}
